# Week 6 — Validation Audit

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/YoyuQre/flyrank_assgn_1/blob/main/work/notebooks/w06_validation_audit.ipynb?flush_cache=true)

This notebook audits the Week-5 Random Forest model (`work/notebooks/w05_model.ipynb`) for **validation rigor, leakage, failure modes, and claim strength**. Everything here is recomputed from the actual dataset and the actual Week-5 pipeline — no metric is typed in from memory.

## 0. Audit objective

The Week-5 model predicts whether a content page is labeled **declining** (`trend_direction == "down"`, a 30-day impression drop beyond the label's ~20% threshold). This audit answers four questions:

1. **How optimistic was the original evaluation?** I reproduce the Week-5 pipeline and compare the original random-split estimate against a grouped-by-client split where the same client never appears in both train and test.
2. **Is there leakage?** I check every feature family for availability at prediction time, label-derivation, future windows, and identifiers.
3. **Where does the model fail?** I show real false positives and false negatives from the honest split.
4. **Are the claims safe?** I rewrite overconfident Week-5 claims in evidence-safe language.

Language note: this notebook uses *observed, measured, directional, decision-support* language throughout. It makes no causal claims and no universal-generalization claims.


## 1. Two paper findings + my methodology questions

Source: `docs/flyrank-seo-research-march-2026.pdf` — *The State of AI-Driven SEO* (FlyRank, March 2026). I selected two findings directly relevant to this project's own methodology choices: the freshness/refresh lever (which the Week-5 model's `days_since_last_update` feature is meant to capture) and the ML appendix's growth model (which uses the same trend-vs-feature-window structure as my pipeline).


### Finding 1 — The freshness multiplier

**What the paper reports.** The paper (Finding #4) reports that the 31–90 day window is the strongest stable freshness band, with a 7.88:1 growth-to-decline ratio. It also reports that mature content (365+ days old) refreshed within 30 days showed a **3.2× health-score boost** (10.7 → 34.5) and **57× more impressions** (71 → 4,039) than comparable untouched pages, in the local active-content sample.

**Methodology question.** One question I would ask when interpreting this finding is whether the "refreshed within 30 days" group and the "untouched" comparison group are selected the same way. Refreshing a page is itself a decision made by an editor who already judged the page worth updating, so the refreshed cohort may be a *better-candidate* group before the refresh — not a random subset that a refresh alone lifted. A useful clarification would be whether the 3.2×/57× figures are reported as an observational difference between pre-existing groups, and how the two cohorts were matched (e.g. by age tier, prior impressions, and topic) so that selection bias is visible. The paper itself flags that this is an observational study ("correlations do not prove causation"); the question above is about how the comparison cohort was constructed within that observational design.


### Finding 2 — What predicts growth (ML appendix)

**What the paper reports.** The exploratory ML appendix reports a logistic regression that separates growing from declining pages with ~71% holdout accuracy on 61.8K active content pieces (an 80/20 split). Content age is the strongest negative coefficient, while days-visible and recent impressions are among the strongest positive signals. The paper explicitly frames these as descriptive indicators, not optimization instructions.

**Methodology question.** One question I would ask when interpreting this finding is whether the 80/20 holdout is a row-level random split and whether pages from the same brand can appear in both folds. If a single brand's pages share hidden characteristics (template, publishing workflow, niche), a row-level random split can make holdout performance look stronger than a brand-level holdout would — the same concern that motivates the grouped split in my own Section 2. A useful clarification would also be whether the label (growing vs declining, defined on the 30-day trend window) overlaps the feature window: if the features are trailing 90-day metrics that include the label's own last-30-day window, the model can partly "read the answer" through those columns. My own audit (Section 3) found exactly this structure in the starter dataset.


## 2. My model under an honest split (before/after)

I reproduce the Week-5 pipeline **verbatim** from `work/notebooks/w05_model.ipynb`: same target, same excluded columns, same missing-value handling, same one-hot encoding, same Random Forest (`n_estimators=200`, `random_state=42`). The only thing I vary between the two evaluations is the **split design**.

Two evaluation numbers exist in the Week-5 record:

* The **original estimate** (accuracy 0.866) came from a single random 80/20 split.
* The **committed Week-5 notebook** already switched to a grouped-by-client split and reports accuracy 0.814.

Both are recomputed below from the actual predictions. The before/after comparison treats the random-split estimate as the "before" and the grouped-by-client split as the "after", because the question of interest is whether the split design inflated the estimate.


In [1]:
import os
import numpy as np
import pandas as pd

from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split, GroupShuffleSplit
from sklearn.metrics import (
    accuracy_score,
    precision_score,
    recall_score,
    f1_score,
    confusion_matrix,
)

def find_csv():
    for path in [
        "data/raw/content_refresh_anonymized.csv",
        "content_refresh_anonymized.csv",
        "/content/content_refresh_anonymized.csv",
    ]:
        if os.path.exists(path):
            return path
    current = os.path.abspath(os.getcwd())
    for _ in range(6):
        probe = os.path.join(current, "data", "raw", "content_refresh_anonymized.csv")
        if os.path.exists(probe):
            return probe
        parent = os.path.dirname(current)
        if parent == current:
            break
        current = parent
    raise FileNotFoundError("Could not locate content_refresh_anonymized.csv")

df = pd.read_csv(find_csv())

# --- Week-5 pipeline reproduced verbatim from w05_model.ipynb ---
data = df.copy()

# Target, exactly as Week 5
y = (data["trend_direction"] == "down").astype(int)

# Excluded columns, exactly as Week 5
drop_cols = ["content_id", "client_id", "trend_direction", "trend_pct"]
X = data.drop(columns=drop_cols)

num_cols = X.select_dtypes(include=["number"]).columns
cat_cols = X.select_dtypes(include=["object"]).columns
X[num_cols] = X[num_cols].fillna(0)
X[cat_cols] = X[cat_cols].fillna("Unknown")
X = pd.get_dummies(X, columns=cat_cols, drop_first=True)

print("Raw data:", df.shape)
print("Feature matrix:", X.shape)
print("Label prevalence (base rate for the 'down' class):", round(float(y.mean()), 4))


Raw data: (30000, 44)
Feature matrix: (30000, 66)
Label prevalence (base rate for the 'down' class): 0.5421


### 2.1 Week-5 baseline (original evaluation)

The original evaluation used a single **random 80/20 train/test split** with stratification (`random_state=42`). Everything else is identical to Week 5. The metrics below are computed from the model's actual predictions on the held-out 20%.


In [2]:
X_train_r, X_test_r, y_train_r, y_test_r = train_test_split(
    X, y, test_size=0.2, random_state=42, stratify=y
)

rf_random = RandomForestClassifier(n_estimators=200, random_state=42)
rf_random.fit(X_train_r, y_train_r)
y_pred_random = rf_random.predict(X_test_r)
y_prob_random = rf_random.predict_proba(X_test_r)[:, 1]

random_results = {
    "Accuracy": accuracy_score(y_test_r, y_pred_random),
    "Precision": precision_score(y_test_r, y_pred_random),
    "Recall": recall_score(y_test_r, y_pred_random),
    "F1": f1_score(y_test_r, y_pred_random),
}

print("Random-split evaluation (Week-5 original estimate):")
print(random_results)
print("Test base rate (label prevalence):", round(float(y_test_r.mean()), 4))


Random-split evaluation (Week-5 original estimate):
{'Accuracy': 0.8663333333333333, 'Precision': 0.8569347319347319, 'Recall': 0.9043665436654367, 'F1': 0.8800119688809096}
Test base rate (label prevalence): 0.542


### 2.2 Honest grouped evaluation (grouped by client)

The honest design groups by `client_id`: all pages belonging to one client go entirely into train or entirely into test, so **no client appears in both**. `client_id` is used only for grouping — it is dropped from the feature matrix and is never a model feature. I verify the separation explicitly with an assertion and report row counts, unique-client counts, and the label base rate for both folds. Preprocessing stays identical to Week 5, so the only difference between the two evaluations is the split.


In [3]:
groups = df["client_id"]
gss = GroupShuffleSplit(n_splits=1, test_size=0.2, random_state=42)
train_idx, test_idx = next(gss.split(X, y, groups=groups))

X_train_g, X_test_g = X.iloc[train_idx], X.iloc[test_idx]
y_train_g, y_test_g = y.iloc[train_idx], y.iloc[test_idx]

train_clients = set(df.iloc[train_idx]["client_id"])
test_clients = set(df.iloc[test_idx]["client_id"])
overlap = train_clients & test_clients

print("Train rows:", len(train_idx), "| unique clients:", len(train_clients))
print("Test rows :", len(test_idx), "| unique clients:", len(test_clients))
print("Client overlap (train & test):", len(overlap))
assert len(overlap) == 0, "Leakage: a client appears in both train and test!"
assert "client_id" not in X.columns, "client_id must not be a model feature"
print("Train base rate:", round(float(y_train_g.mean()), 4),
      "| Test base rate:", round(float(y_test_g.mean()), 4))

rf_grouped = RandomForestClassifier(n_estimators=200, random_state=42)
rf_grouped.fit(X_train_g, y_train_g)
y_pred_grouped = rf_grouped.predict(X_test_g)
y_prob_grouped = rf_grouped.predict_proba(X_test_g)[:, 1]

grouped_results = {
    "Accuracy": accuracy_score(y_test_g, y_pred_grouped),
    "Precision": precision_score(y_test_g, y_pred_grouped),
    "Recall": recall_score(y_test_g, y_pred_grouped),
    "F1": f1_score(y_test_g, y_pred_grouped),
}

print("Honest grouped-by-client evaluation:")
print(grouped_results)
print("Confusion matrix (rows: true, cols: predicted):")
print(confusion_matrix(y_test_g, y_pred_grouped))


Train rows: 23837 | unique clients: 25
Test rows : 6163 | unique clients: 7
Client overlap (train & test): 0
Train base rate: 0.5501 | Test base rate: 0.511


Honest grouped-by-client evaluation:
{'Accuracy': 0.8142138568878793, 'Precision': 0.8049300060864273, 'Recall': 0.8399491902191172, 'F1': 0.822066822066822}
Confusion matrix (rows: true, cols: predicted):
[[2373  641]
 [ 504 2645]]


### 2.3 Before/after comparison

| Evaluation | Accuracy | Precision | Recall | F1 |
| ---------- | -------: | --------: | -----: | --: |
| Week-5 evaluation (random split) | 0.8663 | 0.8569 | 0.9044 | 0.8800 |
| Honest grouped split | 0.8142 | 0.8049 | 0.8399 | 0.8221 |
| Difference (grouped − random) | −0.0521 | −0.0520 | −0.0644 | −0.0579 |

Under the grouped-by-client evaluation, every metric is lower than under the random-split estimate. The gap (≈ 0.05–0.06) is the amount of apparent performance the random split could not support once client groups were kept intact.


In [4]:
comparison = pd.DataFrame({
    "Week-5 evaluation (random split)": random_results,
    "Honest grouped split": grouped_results,
}).T

comparison.loc["Difference (grouped - random)"] = [
    grouped_results[m] - random_results[m] for m in comparison.columns
]
comparison


,Accuracy,Precision,Recall,F1
Week-5 evaluation (random split),0.866333,0.856935,0.904367,0.880012
Honest grouped split,0.814214,0.804930,0.839949,0.822067
Difference (grouped - random),-0.052119,-0.052005,-0.064417,-0.057945


### 2.4 Interpretation

Under the grouped-by-client evaluation, the measured F1 dropped from **0.880** to **0.822** and accuracy from **0.866** to **0.814**. This suggests the original random-split estimate was **optimistic**: a random split lets pages from the same client appear in both folds, and the model can borrow client-level patterns (template, niche, publishing workflow) from train pages to help with test pages. The grouped result provides a more conservative, measured estimate of cross-client performance.

Two caveats, stated plainly:

* The grouped split is still a **one-time split of one dataset**. It is a measured indication of cross-client behavior on this snapshot, not a guarantee for future clients or future data.
* **Base rate matters.** The label is 54% positive, so a trivial "always predict decline" rule scores 0.54 accuracy. Against that reference, 0.814 accuracy on the grouped split is meaningful, but the raw number should never be read without the base rate.


## 3. Leakage audit

I audit the complete Week-5 feature pipeline, not just the final dataframe. For every feature family I check: (1) is it available at prediction time? (2) is it derived from the label? (3) does it contain future information? (4) is it an identifier? (5) was it calculated over the whole dataset before splitting? (6) could it indirectly encode the target? (7) could the same entity appear across train/test?


In [5]:
print("=== Identifier and label-derived checks ===")
check_cols = ["content_id", "client_id", "trend_direction", "trend_pct", "is_declining_label"]
for col in check_cols:
    print(f"{'LEAK FOUND' if col in X.columns else 'ok           '}  {col:22s} in feature set: {col in X.columns}")

assert not any(c in X.columns for c in check_cols), "leakage check failed"
assert "is_declining_label" not in X.columns, "the label itself must not be a feature"
print("\nNo label-derived or identifier columns are present in the final feature set.")

print("\n=== Leak-detection sanity check ===")
print("The harness must be able to detect leakage when it exists. Retrain with trend_pct")
print("(the label's source column) added back, everything else identical:")
X_leak = data.copy().drop(columns=["content_id", "client_id", "trend_direction"])
nl = X_leak.select_dtypes(include=["number"]).columns
cl = X_leak.select_dtypes(include=["object"]).columns
X_leak[nl] = X_leak[nl].fillna(0)
X_leak[cl] = X_leak[cl].fillna("Unknown")
X_leak = pd.get_dummies(X_leak, columns=cl, drop_first=True)

Xtr_l, Xte_l, ytr_l, yte_l = train_test_split(
    X_leak, y, test_size=0.2, random_state=42, stratify=y
)
rf_leak = RandomForestClassifier(n_estimators=200, random_state=42)
rf_leak.fit(Xtr_l, ytr_l)
leak_acc = accuracy_score(yte_l, rf_leak.predict(Xte_l))
print(f"Accuracy with trend_pct as a feature: {leak_acc:.4f}  (vs {random_results['Accuracy']:.4f} without)")
print("The jump toward 1.0 confirms the harness detects leakage, and explains why trend_direction")
print("and trend_pct must never be features.")


=== Identifier and label-derived checks ===
ok             content_id             in feature set: False
ok             client_id              in feature set: False
ok             trend_direction        in feature set: False
ok             trend_pct              in feature set: False
ok             is_declining_label     in feature set: False

No label-derived or identifier columns are present in the final feature set.

=== Leak-detection sanity check ===
The harness must be able to detect leakage when it exists. Retrain with trend_pct
(the label's source column) added back, everything else identical:


Accuracy with trend_pct as a feature: 0.9997  (vs 0.8663 without)
The jump toward 1.0 confirms the harness detects leakage, and explains why trend_direction
and trend_pct must never be features.


### 3.1 Feature availability

All features in the Week-5 matrix come from the **trailing-90-day snapshot** of the dataset: search-performance aggregates (90-day and 30-day windows), engagement metrics, content metadata, and threshold-based tiers. The dataset is a static export, so every column is known at the moment a prediction would be made; there is no separate future window in the starter data. The remaining audit questions are therefore (a) identifiers, (b) label-derived columns, and (c) whether the 30-day trend-window columns can reconstruct the label.

### 3.2 Identifier and label-derived feature checks

* `content_id` and `client_id` are pseudonymous identifiers — grouping/joins only, never features. Verified excluded above.
* The label is `is_declining = (trend_direction == "down")`, and `trend_direction` is computed from `trend_pct = (impressions_last_30d - impressions_prev_30d) / impressions_prev_30d`. Therefore `trend_direction` and `trend_pct` are label siblings and are excluded. Verified excluded above.
* The one **gray zone**: `impressions_last_30d` and `impressions_prev_30d` are kept as features (they are legitimate raw inputs available at prediction time), but the label is a deterministic function of their ratio. The model can therefore reconstruct much of the label from these two columns. This is **not classic leakage** (the columns are not the label and are available at prediction time), but it does mean the high importance of these features reflects label reconstruction, and part of the accuracy ceiling is the model restating the trend. The Week-5 pipeline is preserved as-is; the caveat is disclosed here and reflected in Section 5.

### 3.3 Leakage audit table

| Feature group | Members | Available at prediction time? | Label-derived? | Future info? | Identifier? | Decision | Reason |
| ------------- | ------- | ----------------------------- | -------------- | ------------ | ----------- | -------- | ------ |
| Identifiers | content_id, client_id | — | No | No | Yes | **Exclude** | Pseudonymous IDs; grouping/joins only. Using them as features enables row memorization and leaks group structure. |
| Label source | trend_direction, trend_pct | — | Yes | — | No | **Exclude** | `is_declining_label` is derived from `trend_direction`, which is computed from `trend_pct`. Both are the target's own columns. |
| 90-day aggregates | impressions_90d, clicks_90d, pageviews_90d, sessions_90d, users_90d, engaged_sessions_90d, ai_sessions_90d, scroll_events_90d, days_with_impressions, days_with_sessions | Yes | No | No | No | **Include** | Historical trailing-90d window ending at export time; known before prediction. |
| 30-day trend windows | impressions_last_30d, clicks_last_30d, sessions_last_30d, impressions_prev_30d, clicks_prev_30d, sessions_prev_30d | Yes | Partially (the label is a function of the last-vs-prev impression ratio) | No | No | **Needs review** | Legitimate raw inputs, but `impressions_last_30d` vs `impressions_prev_30d` nearly reconstruct `trend_pct`, the label's source. Kept to preserve the Week-5 pipeline; treated as label-reconstructive, not independent signal. |
| Keyword context | search_volume, competition, competition_level, cpc | Yes | No | No | No | **Include** | Keyword metadata captured with the snapshot. |
| Content properties | word_count, char_count, content_age_days, days_since_last_update, provider_used, model_used, age_tier, age_tier_order, freshness_tier, word_count_tier, char_count_tier | Yes | No | No | No | **Include** | Static content metadata / deterministic tiers, known before prediction. |
| Derived rates | ctr, avg_position, engagement_rate, scroll_rate, ai_traffic_pct | Yes | No | No | No | **Include** | Computed from trailing-90d aggregates; available before prediction. Note: `avg_position = 0` means "no data" (1,205 rows), not rank zero. |
| Tier columns | impression_tier, position_tier | Yes | No | No | No | **Include** | Deterministic recodings of 90-day aggregates and position; no label information. |

**Verdict.** No label-derived or identifier columns are present in the final feature set, and the label itself is not a feature. The 30-day trend-window columns are the only family that needs review: they are legitimate at prediction time, but they make part of the task near-tautological, so the headline accuracy is best read as "given the trend windows, classify the trend" rather than "detect content decline from independent signals".


## 4. Real failure examples

Examples below come from the **honest grouped-by-client evaluation** (Section 2.2). Nothing is fabricated. `content_id` and `client_id` in this dataset are already pseudonymized (the data dictionary confirms they are `content_`/`client_` + hex), so showing them does not expose real URLs, domains, or client names. Predictions are the model's `>= 0.5` decisions; `prob_decline` is the predicted probability of the "down" class.


In [6]:
fail = df.iloc[test_idx][["content_id", "client_id"]].copy()
fail["y_true"] = y_test_g.values
fail["y_pred"] = y_pred_grouped
fail["prob_decline"] = np.round(y_prob_grouped, 3)

show_cols = [
    "impressions_last_30d", "impressions_prev_30d", "impressions_90d",
    "clicks_last_30d", "ctr", "avg_position",
    "content_age_days", "days_since_last_update", "word_count",
]
for col in show_cols:
    fail[col] = df.iloc[test_idx][col].values

false_positives = fail[(fail.y_pred == 1) & (fail.y_true == 0)].sort_values(
    "prob_decline", ascending=False
)
false_negatives = fail[(fail.y_pred == 0) & (fail.y_true == 1)].sort_values(
    "prob_decline", ascending=True
)

print("False positives (predicted decline, actual not declining):", len(false_positives))
print("False negatives (missed decline, actual declining):", len(false_negatives))


False positives (predicted decline, actual not declining): 641
False negatives (missed decline, actual declining): 504


In [7]:
print("False positives - sample (highest 'decline' probability, actually NOT declining)")
false_positives[["content_id", "client_id", "y_true", "y_pred", "prob_decline"] + show_cols].head(6)


False positives - sample (highest 'decline' probability, actually NOT declining)


,content_id,client_id,y_true,y_pred,prob_decline,impressions_last_30d,impressions_prev_30d,impressions_90d,clicks_last_30d,ctr,avg_position,content_age_days,days_since_last_update,word_count
2357,content_8f1409b2674e,client_8527a891e2,0,1,0.910,32,39,209,0,0.00,20.0,271,104,1450.0
20736,content_41baf0722ad9,client_8527a891e2,0,1,0.870,276,322,3115,0,0.00,12.8,275,104,1596.0
12332,content_4d9f36001f06,client_8527a891e2,0,1,0.865,268,323,3369,0,0.03,13.2,275,104,1643.0
10080,content_35d63627bf3e,client_8527a891e2,0,1,0.855,364,335,1525,0,0.00,32.6,238,103,1592.0
19926,content_b84a5df88090,client_f369cb89fc,0,1,0.825,732,866,3699,0,0.03,8.1,95,20,2977.0
1517,content_816d77e36e14,client_8527a891e2,0,1,0.820,38,47,208,0,0.48,25.3,275,104,1550.0


In [8]:
print("False negatives - sample (highest 'not declining' confidence, actually declining)")
false_negatives[["content_id", "client_id", "y_true", "y_pred", "prob_decline"] + show_cols].head(6)


False negatives - sample (highest 'not declining' confidence, actually declining)


,content_id,client_id,y_true,y_pred,prob_decline,impressions_last_30d,impressions_prev_30d,impressions_90d,clicks_last_30d,ctr,avg_position,content_age_days,days_since_last_update,word_count
20902,content_d663eacf9688,client_4e07408562,1,0,0.195,351,509,1326,0,0.08,24.9,537,7,2760.0
1010,content_8ff857ae67d0,client_e629fa6598,1,0,0.200,5459,7196,21103,47,1.11,4.2,460,22,NaN
9233,content_1d2db6db4635,client_4e07408562,1,0,0.210,491,707,2085,0,0.24,5.9,445,7,2504.0
7744,content_b26ac45aca5d,client_4e07408562,1,0,0.225,343,821,2916,1,0.14,36.8,445,7,2646.0
24864,content_8ab9f1033843,client_e629fa6598,1,0,0.235,407,562,1265,0,0.00,37.2,460,22,NaN
10503,content_5e1febd65aa2,client_4e07408562,1,0,0.235,544,746,2091,1,0.29,7.6,420,7,2970.0


### 4.1 False positives

In the sampled false positives, I observed pages whose 30-day impressions dipped **within the stable band** — real but milder than the label's ~20% decline threshold — while their engagement and visibility signals were weak. For example, one false positive (predicted decline at 0.91 probability) had 32 impressions in the last 30 days vs 39 in the previous 30 days (−18%, above the −20% label threshold so not labeled declining), zero clicks, `ctr = 0`, average position 20.0, and no update in 104 days. The model's decision boundary sits below the label's. Several other false positives carried **deep positions** (avg_position 20–33) or were **long-unupdated** (104 days), so the model leaned on a "stale + weakly visible" profile even where impressions were roughly flat.

### 4.2 False negatives

The false-negative examples suggest the model was misled by **strong-looking profiles that still declined**. In the sampled false negatives, pages dropped more than 20% in last-30-day impressions (e.g. 5,459 → 7,196, −24%; 351 → 509, −31%; 343 → 821, −58%) yet the model predicted not-declining. These pages were recently updated (days_since_last_update 7–22), several sat on page 1 (avg_position 4–8) with healthy CTR (0.24–1.11) and high impression volume. In other words, the model's freshness/visibility signals pointed "safe" while the actual trend was down.

### 4.3 What I observed

These examples are diagnostic rather than sufficient to establish a general causal explanation. The patterns are consistent with the model's top features (the 30-day impression windows, position, content age): it reconstructs steep, clearly-labeled trends well, but it struggles at the boundary (mild dips below the label threshold read as decline) and when a page profile looks healthy while the trend is down. A decision-support reader should treat high-confidence flags on stale, weakly-visible pages as plausible candidates, and should not treat a "safe" prediction on a fresh, well-ranked page as proof the page is not declining.


## 5. Claim rewrite

The claims below are taken **verbatim** from `work/notebooks/w05_model.ipynb` (and the earlier Week-6 draft), then rewritten in evidence-safe language. Nothing is invented.

### 5.1 Claims that go too far

| Original claim | Why it goes too far | Evidence-safe rewrite |
| -------------- | ------------------- | --------------------- |
| "The Random Forest model performed substantially better than the rule-based baseline, indicating that combining multiple signals is more effective than relying on a few fixed thresholds." (w05, §4) | "Indicating that combining multiple signals is more effective" turns a measured result on one dataset into a general mechanism claim. | Under the grouped-by-client evaluation, the Random Forest measured higher accuracy (0.814 vs 0.527) and F1 (0.822 vs 0.629) than the rule baseline on this dataset; this is a measured, dataset-specific comparison. |
| "It is also robust to noisy features and provides feature importance scores, making it useful for both prediction and interpretation." (w05, §1) | "Robust to noisy features" is asserted as a property of the model rather than something measured in this project. | Random Forest provides per-feature importance scores from the fitted model; the stability of those importances across re-fits was not measured in this project. |
| "the grouped split ... results are more representative of how the model would perform on new, unseen clients." (Week-6 draft) | "More representative ... on new, unseen clients" implies cross-client generalization beyond one snapshot. | Performance under the grouped-by-client evaluation provides a measured indication of cross-client behavior on this dataset; it is not a guarantee for unseen clients or future data. |
| "This suggests that content decline is influenced by multiple interacting factors" (w05, §4) | "Influenced by" is a causal reading of feature-importance and error analysis. | Feature-importance inspection shows the fitted model weighted several feature families (impression history, position, content age, engagement); this describes model behavior on this dataset, not a causal explanation of decline. |
| "The Random Forest model accurately predicts which content pages will decline in search performance." (Week-6 draft) | "Will decline" promises future behavior; "accurately predicts" overstates a single-split measurement. | In this dataset, the Random Forest measured strong classification performance (F1 0.822 under the grouped split) for identifying pages labeled as declining, where the label is a 30-day impression decline. The model is decision support for prioritizing content review; downstream business impact was not measured. |

### 5.2 Evidence-safe rewrites

The rewrites above stick to the safe vocabulary: **observed, measured, directional, decision-support**. Where the original made a general claim, the rewrite says what was actually measured, on which dataset, and under which split. Where the original implied causality, the rewrite explicitly labels the analysis as descriptive.


## 6. Self-check

Final checklist for this audit notebook:

- [x] Two paper findings identified (freshness multiplier; ML appendix growth model)
- [x] Methodology question written for finding 1
- [x] Methodology question written for finding 2
- [x] Week-5 baseline reproduced from code (random split -> accuracy 0.866; grouped split -> accuracy 0.814)
- [x] Honest grouped-by-client split implemented (GroupShuffleSplit on `client_id`)
- [x] Train/test group overlap checked (asserted zero overlap: 25 train clients, 7 test clients)
- [x] Before/after metrics shown in a comparison table with differences
- [x] Leakage audit completed (identifier + label-source checks, feature-family table, leak-detection sanity test)
- [x] Identifier/label-derived features excluded (`content_id`, `client_id`, `trend_direction`, `trend_pct` verified absent)
- [x] Real false-positive examples shown (641 available, sample displayed)
- [x] Real false-negative examples shown (504 available, sample displayed)
- [x] Claims rewritten using evidence-safe language (Section 5)
- [x] No unsupported causal/generalization claims (interpretations explicitly marked descriptive/diagnostic)
- [x] Notebook executed successfully (kernel run from top to bottom)
- [x] Notebook saved with outputs

### Conclusion — what changed in my understanding

Running this audit changed how I read the Week-5 result in three ways.

1. **The original estimate was optimistic.** The random-split estimate (F1 0.880, accuracy 0.866) was not reproducible under a grouped-by-client split (F1 0.822, accuracy 0.814). The gap of roughly 0.05–0.06 is the portion of the headline score that depended on client groups appearing in both folds. The honest, measured numbers are 0.814 accuracy / 0.822 F1.

2. **The label is nearly reconstructable from two kept features.** Because `is_declining` is a function of `impressions_last_30d` vs `impressions_prev_30d`, the model's top features effectively re-derive the label. This is not classic leakage, but it caps the amount of "independent signal" the model can claim — the headline accuracy is partly the model restating the trend, not detecting decline from unrelated signals.

3. **I will report measurements, not guarantees.** The claims I am now comfortable making describe what was observed on this snapshot under a specific split. Anything about future clients, future months, or *why* pages decline is hypothesis and decision support, not a finding.
